addons/sale_coupon/models/sale_order.py

```diff
@@ -157,7 +157,16 @@ def _get_reward_values_discount(self, program):
             }]
         reward_dict = {}
         lines = self._get_paid_order_lines()
-        amount_total = sum(self._get_base_order_lines(program).mapped('price_subtotal'))
+        base_lines = self._get_base_order_lines(program)
+
+        base_lines_price_include = base_lines.filtered(
+            lambda l: any(l.tax_id.mapped(lambda tax_id: tax_id.price_include)))
+        base_lines_price_exclude = base_lines - base_lines_price_include
+
+        # Calculating tax included prices and tax excluded prices separately
+        amount_total = 0
+        amount_total += sum(base_lines_price_include.mapped('price_total'))
+        amount_total += sum(base_lines_price_exclude.mapped('price_subtotal'))
```

Comment:
This seems a bit over complicated:
- ```l.tax_id.mapped(lambda tax_id: tax_id.price_include)``` => ```l.tax_id.mapped('price_include')```
- initializing variable to 0 to just add something the next line is just the same as initializing the variable with what you will add
```a=0; a+=2;``` => ```a=2```

At the end, this is just a one liner (make it 2 for nice indentation), which is way easier to read and understand

```suggestion
        amount_total = sum([any(line.tax_id.mapped('price_include')) and line.price_total or line.price_subtotal
                            for line in self._get_base_order_lines(program)])
```